pkg/dockerregistry/server/quotarestrictedblobstore.go

```diff
@@ -122,7 +123,51 @@ func (bw *quotaRestrictedBlobWriter) Commit(ctx context.Context, provisional dis
 
 // admitBlobWrite checks whether the blob does not exceed image quota, if set. Returns
 // ErrAccessDenied error if the quota is exceeded.
-func admitBlobWrite(ctx context.Context, repo *repository) error {
+func admitBlobWrite(ctx context.Context, repo *repository, size int64) error {
+	if size < 1 {
+		return nil
+	}
+
+	err := admitLimitRanges(ctx, repo, size)
+	if err != nil {
+		return err
+	}
+
+	err = admitQuotas(ctx, repo)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// admitLimitRanges checks the blob against any established limit ranges.
+func admitLimitRanges(ctx context.Context, repo *repository, size int64) error {
+	limitranges, err := repo.limitClient.LimitRanges(repo.namespace).List(kapi.ListOptions{})
+	if err != nil {
+		if kerrors.IsForbidden(err) {
+			context.GetLogger(ctx).Warnf("Cannot list limitranges because of outdated cluster roles: %v", err)
```

Comment:
The other reason I don't like forcing this - what if the admin has custom
roles and they want to reconcile them?  Their registry is blocked until
they grant this permission?

I'd be in favor of a check somewhere else that warns / notifies the admin
that their registry can't enforce quota.  I'd be in favor of having a
boolean flag on the registry that allows the registry to hard fail on
quota.  I'm not comfortable breaking the registry on upgrade for a net-new
feature.

On Tue, Mar 22, 2016 at 9:14 PM, Jordan Liggitt <notifications@github.com>
wrote:

> In pkg/dockerregistry/server/quotarestrictedblobstore.go
> <https://github.com/openshift/origin/pull/7989#discussion_r57096544>:
>
> > +	}
> > +
> > +	err = admitQuotas(ctx, repo)
> > +	if err != nil {
> > +		return err
> > +	}
> > +
> > +	return nil
> > +}
> > +
> > +// admitLimitRanges checks the blob against any established limit ranges.
> > +func admitLimitRanges(ctx context.Context, repo *repository, size int64) error {
> > +	limitranges, err := repo.limitClient.LimitRanges(repo.namespace).List(kapi.ListOptions{})
> > +	if err != nil {
> > +		if kerrors.IsForbidden(err) {
> > +			context.GetLogger(ctx).Warnf("Cannot list limitranges because of outdated cluster roles: %v", err)
>
> I would not expect quota to be involved at all in read-only registry paths
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7989/files/aee1f1537969521a96e027cec8461be7689d4436..e8e272d1d258f7b0f5941504ca83cb792bfc2098#r57096544>
>